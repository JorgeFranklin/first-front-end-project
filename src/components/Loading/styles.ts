import styled, { css, keyframes } from 'styled-components'

const rotate = keyframes`
  0% { transform: rotate(0deg)  }
  100% { transform: rotate(360deg)  }
`

const openAndClose = keyframes`
  0% { height: 2rem; width: 4rem }
  25% { height: 1rem; width: 2rem }
  50% { height: 1rem; width: 1rem }
  75% { height: 1rem; width: 2rem }
  100% { height: 2rem; width: 4rem }
`

export const Wrapper = styled.div`
  height: 4.5rem;
  width: 4.5rem;
`

type CircleProps = {
  color?: 'primary' | 'secondary'
  bgColor?: 'black' | 'white'
}

export const Circle = styled.div<CircleProps>`
  ${({ theme, color, bgColor }) => css`
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    border-radius: 50%;
    border: ${theme.colors[color!]} 0.5rem solid;
    width: 100%;
    height: 100%;
    animation: ${rotate} 1s linear infinite;

    ::before {
      content: '';
      position: absolute;
      background-color: ${bgColor === 'black'
        ? theme.colors.mainBg
        : theme.colors.white};
      top: -0.7rem;
      border-radius: 50%;
      animation: ${openAndClose} 3s infinite reverse backwards;
    }
  `}
`
