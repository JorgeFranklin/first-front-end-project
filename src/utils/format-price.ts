export default function formatPrice(price: number | bigint): string {
  if (price <= 0) return 'FREE'

  const formattedPrice = new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD'
  }).format(price)

  return formattedPrice
}
