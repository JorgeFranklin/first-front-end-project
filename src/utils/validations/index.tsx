import { UsersPermissionsRegisterInput } from 'graphql/generated/globalTypes'
import Joi from 'joi'

const fieldsValidation = {
  username: Joi.string().min(5).required(),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),
  password: Joi.string().required(),
  confirm_password: Joi.string()
    .valid(Joi.ref('password'))
    .required()
    .messages({ 'any.only': 'confirm password does not match with password' })
}

export type FieldErrors = {
  [field: string]: string
}

export function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path?.join('.')] = err.message
    })
  }

  return errors
}

export function signUpValidate(values: UsersPermissionsRegisterInput) {
  const schema = Joi.object(fieldsValidation)
  const objError = schema.validate(values, { abortEarly: false })

  return getFieldErrors(objError)
}

type SignInValues = Omit<UsersPermissionsRegisterInput, 'username'>

export function signInValidate(values: SignInValues) {
  const { email, password } = fieldsValidation
  const schema = Joi.object({ email, password })

  const objError = schema.validate(values, { abortEarly: false })

  return getFieldErrors(objError)
}

type ForgotValidate = Pick<UsersPermissionsRegisterInput, 'email'>

export function forgotValidate(values: ForgotValidate) {
  const { email } = fieldsValidation
  const schema = Joi.object({ email })

  const objError = schema.validate(values, { abortEarly: false })

  return getFieldErrors(objError)
}

type ResetValidate = {
  password: string
  confirm_password: string
}

export function resetValidate(values: ResetValidate) {
  const { password, confirm_password } = fieldsValidation
  const schema = Joi.object({ password, confirm_password })

  const objError = schema.validate(values, { abortEarly: false })

  return getFieldErrors(objError)
}
