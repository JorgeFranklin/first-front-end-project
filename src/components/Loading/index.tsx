import * as S from './styles'

export type LoadingProps = {
  color?: 'primary' | 'secondary'
  bgColor?: 'black' | 'white'
}

const Loading = ({ color = 'primary', bgColor = 'black' }: LoadingProps) => (
  <S.Wrapper>
    <S.Circle color={color} bgColor={bgColor} />
  </S.Wrapper>
)

export default Loading
