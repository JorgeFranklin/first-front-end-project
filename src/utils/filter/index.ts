import { ItemProps } from 'components/ExploreSidebar'
import { ParsedUrlQueryInput } from 'querystring'

type ParseToWhereArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToWhere = ({
  queryString,
  filterItems
}: ParseToWhereArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString)
    .filter((key) => key !== 'sort')
    .forEach((key) => {
      const item = filterItems?.find((item) => item.name === key)
      const isCheckbox = item?.type === 'checkbox'

      obj[key] = isCheckbox
        ? { name_contains: queryString[key] }
        : queryString[key]
    })

  return obj
}

type ParseToFilterArgs = {
  queryString: ParsedUrlQueryInput
  filterItems: Pick<ItemProps, 'type' | 'name'>[]
}

export const parseQueryStringToFilter = ({
  queryString,
  filterItems
}: ParseToFilterArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any
  const obj: any = {}

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key)
    const isCheckbox = item?.type === 'checkbox'
    const isArray = Array.isArray(queryString[key])

    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key]
  })

  return obj
}
