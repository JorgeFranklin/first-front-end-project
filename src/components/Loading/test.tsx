import { render } from 'utils/test-utils'

import Loading from '.'

describe('<Loading />', () => {
  it('should render loading', () => {
    const { container } = render(<Loading />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
