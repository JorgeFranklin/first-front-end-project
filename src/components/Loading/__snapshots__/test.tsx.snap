// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render loading 1`] = `
.c0 {
  height: 4.5rem;
  width: 4.5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  border-radius: 50%;
  border: #F231A5 0.5rem solid;
  width: 100%;
  height: 100%;
  -webkit-animation: bhnNqB 1s linear infinite;
  animation: bhnNqB 1s linear infinite;
}

.c1::before {
  content: '';
  position: absolute;
  background-color: #06092B;
  top: -0.7rem;
  border-radius: 50%;
  -webkit-animation: iDRrLr 3s infinite reverse backwards;
  animation: iDRrLr 3s infinite reverse backwards;
}

<div
  class="c0"
>
  <div
    class="c1"
    color="primary"
  />
</div>
`;
